### PR TITLE
Ipfs integration

### PR DESCRIPTION
**Pull request by dvc94ch**
Adds dag protobuf support for ipfs

### CODE REVIEW

File: base/src/error.rs [hunk 1]
```diff
@@ -50,31 +50,31 @@ impl From<core::convert::Infallible> for IpldError {
 }
 
 /// Block error.
-#[derive(Debug, Fail)]
+#[derive(Debug, Error)]
 pub enum BlockError {
     /// Block exceeds MAX_BLOCK_SIZE.
-    #[fail(display = "Block size {} exceeds MAX_BLOCK_SIZE.", _0)]
+    #[error("Block size {0} exceeds MAX_BLOCK_SIZE.")]
     BlockToLarge(usize),
```
Comment (aphelionz) [inline base/src/error.rs]: BlockTooLarge*

File: Cargo.toml [hunk 2]
```diff
@@ -12,12 +12,13 @@ byteorder = "1.3.4"
...
 dag-cbor = { path = "dag-cbor" }
 dag-cbor-derive = { path = "dag-cbor-derive" }
-failure = "0.1.7"
+dag-pb = { path = "dag-pb", optional = true }
 futures = "0.3.4"
 libipld-base = { path = "base" }
 libipld-macro = { path = "macro" }
 multibase = "0.8.0"
 multihash = "0.10.1"
+thiserror = "1.0.11"
```
Comment (koivunej) [inline Cargo.toml:21]: I wonder if this could be "1.0" or is there something behind pinning a "past" 1.0 version?

File: dag-cbor/Cargo.toml [hunk 3]
```diff
@@ -7,8 +7,8 @@ license = "MIT OR Apache-2.0"
 
 [dependencies]
 byteorder = "1.3.4"
-failure = "0.1.7"
 libipld-base = { path = "../base" }
+thiserror = "1.0.11"
```
Comment (koivunej) [inline dag-cbor/Cargo.toml:11]: Same pinning here.

File: dag-pb/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "dag-pb"
+version = "0.1.0"
+authors = ["David Craven <david@craven.ch>"]
+edition = "2018"
+
+[dependencies]
+libipld-base = { path = "../base" }
+prost = "0.6.1"
+thiserror = "1.0.11"
```
Comment (koivunej) [inline dag-pb/Cargo.toml:10]: And here.

File: dag-pb/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+use core::convert::TryInto;
+use libipld_base::cid;
+use libipld_base::codec::Codec;
+use libipld_base::error::{BlockError, IpldError};
+use libipld_base::ipld::Ipld;
+use thiserror::Error;
+
+mod codec;
+
+/// Protobuf codec.
```
Comment (koivunej) [inline dag-pb/src/lib.rs:12]: Could link to spec here https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-pb.md

File: dag-pb/src/dag_pb.proto [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+  uint64 Tsize = 3;
+}
+
+// An IPFS MerkleDAG Node
+message PBNode {
+  // refs to other objects
+  repeated PBLink Links = 2;
+  // opaque user data
+  bytes Data = 1;
+}
```
Comment (koivunej) [inline dag-pb/src/dag_pb.proto:22]: This file itself has been updated in the spec with different comments: https://github.com/ipld/specs/blob/master/block-layer/codecs/dag-pb.md

**Review by koivunej — COMMENTED** [review]
This is looking good.

File: Cargo.toml (hunk 2 above)
Comment (dvc94ch) [reply Cargo.toml:21]: patch versions should be compatible. If the crates semver correctly, the best practice imo is to use the version that it was compiled against, like `cargo update` or `cargo upgrade` do.

Comment (koivunej) [reply Cargo.toml:21]: Oki, I'll resolve these then.

**Review by vmx — COMMENTED** [review]
Looks good! Running `cargo fmt` and it should be ready to go.

**Review by vmx — APPROVED** [review]
Once `cargo fmt` is applied it's ready to be merged.

**Pull request merged**